### PR TITLE
Fix convex hull issue by passing compare strategies

### PR DESCRIPTION
**Pull request by vissarion**
This PR propose a fix for https://github.com/boostorg/geometry/issues/1158

First, it enables passing (umbrella) strategies to less policy. Then different (umbrella) strategies can parameterize the less policy by different compare strategy. 

For, example floating point comparison in convex hull strategies does *not* need to be done approximately. Relate strategies still use the approximate comparison of fp numbers (`math::equals()` function).

Side note: I am not sure if approximate comparison of numbers is needed for relate strategy. I just left it as is in order for the tests to pass, but I think it needs some investigation. Related issue https://github.com/boostorg/geometry/issues/1148 

Next steps (to be implemented in new PRs)
- do the same for `greater` and `equal_to`
- move code from `strategies/compare.hpp` to `strategies/cartesian/compare.hpp` (if I understand the strategies code structure correctly, @awulkiew please correct me if I am wrong here)

### CODE REVIEW

File: include/boost/geometry/algorithms/detail/convex_hull/graham_andrew.hpp [hunk 1]
```diff
@@ -225,6 +224,7 @@ class graham_andrew
             point_type const& left, point_type const& right,
             SideStrategy const& side)
     {
+        //std::cout << "add to hull" << std::endl;
```
Comment (barendgehrels) [inline include/boost/geometry/algorithms/detail/convex_hull/graham_andrew.hpp:227]: Is this a left over?

File: include/boost/geometry/algorithms/detail/disjoint/multipoint_geometry.hpp [hunk 2]
```diff
@@ -63,12 +63,13 @@ namespace detail { namespace disjoint
 class multipoint_multipoint
 {
 private:
-    template <typename Iterator, typename CSTag>
+    template <typename Iterator, typename Strategy>
     class unary_not_disjoint_predicate
-        : geometry::less<void, -1, CSTag>
+        : geometry::less<void, -1, Strategy>
```
Comment (barendgehrels) [inline include/boost/geometry/algorithms/detail/disjoint/multipoint_geometry.hpp:68]: Can we call it maybe: `CompareStrategy` ?

File: include/boost/geometry/algorithms/detail/disjoint/multipoint_geometry.hpp [hunk 3]
```diff
@@ -63,12 +63,13 @@ namespace detail { namespace disjoint
...
     {
     private:
-        typedef geometry::less<void, -1, CSTag> base_type;
+        //TODO: pass strategy
```
Comment (barendgehrels) [inline include/boost/geometry/algorithms/detail/disjoint/multipoint_geometry.hpp:71]: Is this still TODO or already done? Because it is passed now. Or do you want to pass an instance? If so, can you rephrase the comment?

File: include/boost/geometry/algorithms/detail/disjoint/multipoint_geometry.hpp [hunk 4]
```diff
@@ -108,7 +109,7 @@ class multipoint_multipoint
         using predicate_type = unary_not_disjoint_predicate
             <
                 typename std::vector<point1_type>::const_iterator,
-                cs_tag
+                Strategy
```
Comment (barendgehrels) [inline include/boost/geometry/algorithms/detail/disjoint/multipoint_geometry.hpp:112]: This (for example) looks like a breaking change. Can you point if there is something which could (potentially) cause non compilations for existing users?

File: include/boost/geometry/algorithms/detail/relate/point_point.hpp [hunk 5]
```diff
@@ -223,7 +222,8 @@ struct multipoint_multipoint
     {
         // sort points from the 1 MPt
         typedef typename geometry::point_type<SortedMultiPoint>::type point_type;
-        typedef geometry::less<void, -1, CSTag> less_type;
+        //TODO: pass strategy
+        typedef geometry::less<void, -1, Strategy> less_type;
```
Comment (barendgehrels) [inline include/boost/geometry/algorithms/detail/relate/point_point.hpp:226]: I'm a bit worried that `geometry::less` is changed. This really looks like a breaking change to me...

So it fixes convex hull (good!) but it needs also changes in many other operations like `relate` and `is_valid` etc...

**Review by barendgehrels — CHANGES_REQUESTED** [review]
As mentioned in my comments, this looks like breaking...

Can it be solved in a different way too? A way does not change `less` itself?

File: include/boost/geometry/policies/compare.hpp [hunk 6]
```diff
@@ -41,7 +42,7 @@ template
 <
     typename Point = void,
     int Dimension = -1,
-    typename CSTag = void
+    typename Strategy = void
```
Comment (barendgehrels) [inline include/boost/geometry/policies/compare.hpp:45]: In summary, this is the main thing I'm worried about.

**Comment by barendgehrels** [comment]
Copying your description, maybe just:
* `less_with_compare_strategy` 
* `greater_with_compare_strategy`
* etc
Though these names are a bit long.

Alternatively the `less` could be meta-templated such that it accepts either `CSTag` or a `CompareStrategy`.

File: include/boost/geometry/algorithms/detail/convex_hull/graham_andrew.hpp (hunk 1 above)
Comment (vissarion) [reply include/boost/geometry/algorithms/detail/convex_hull/graham_andrew.hpp:227]: Yes, will be removed (together with a few other comments).

File: include/boost/geometry/algorithms/detail/disjoint/multipoint_geometry.hpp (hunk 2 above)
Comment (vissarion) [reply include/boost/geometry/algorithms/detail/disjoint/multipoint_geometry.hpp:68]: I do not know, I kept it that way since this is the name used by apply function (below) that actually pass that type.

File: include/boost/geometry/algorithms/detail/disjoint/multipoint_geometry.hpp (hunk 3 above)
Comment (vissarion) [reply include/boost/geometry/algorithms/detail/disjoint/multipoint_geometry.hpp:71]: This is done, removed now.

File: include/boost/geometry/algorithms/detail/disjoint/multipoint_geometry.hpp (hunk 4 above)
Comment (vissarion) [reply include/boost/geometry/algorithms/detail/disjoint/multipoint_geometry.hpp:112]: All strategies (apart from convex hull strategy) calls the old code at the end so AFAIU there is no issue at this point. But there are breaking changes in this PR (see below).

File: include/boost/geometry/algorithms/detail/relate/point_point.hpp (hunk 5 above)
Comment (vissarion) [reply include/boost/geometry/algorithms/detail/relate/point_point.hpp:226]: For everything apart from convex hull the same strategy is called. E.g. for cartesian it is `cartesian<ComparePolicy, EqualsPolicy, -1>` which is as the old one https://github.com/boostorg/geometry/blob/develop/include/boost/geometry/strategies/compare.hpp#L165 (also called from default compare strategy) with the only difference that now it has the EqualsPolicy parameter which for all algorithms is as before (i.e. `strategy::compare::equals_epsilon` see `strategies/relate/cartesian.hpp` but for convex hull it is `strategy::compare::equals_exact` see `strategies/convex_hull/cartesian.hpp`). This change eventually change only convex hull strategies.

File: include/boost/geometry/policies/compare.hpp (hunk 6 above)
Comment (vissarion) [reply include/boost/geometry/policies/compare.hpp:45]: See above.

**Comment by vissarion** [comment]
First, many thanks for the review!

> Copying your description, maybe just:
> 
>     * `less_with_compare_strategy`
> 
>     * `greater_with_compare_strategy`
> 
>     * etc
>       Though these names are a bit long.
> 
> 
> Alternatively the `less` could be meta-templated such that it accepts either `CSTag` or a `CompareStrategy`.

The breaking change I see is in the case some user define a less operator with a CSTag template parameter. I added specializations to support this case too. Please let me know what you think after these changes. 

Side comment, I think it is more correct (according to the library design) to pass a strategy to less than a cs_tag.

**Comment by awulkiew** [comment]
Strategies are for CS-specific computation. AFAIU you want to compare coordinates strictly instead with epsilon in an algorithm. So the differentiation is made based on algorithm, not CS. This probably should not be a strategy but a policy (custom less policy) used only in convex_hull.

Consider this. The most general umbrella strategy can be passed to all algorithms, including `relate` and `convex_hull`. Which compare strategy should be returned by it?

**Comment by awulkiew** [comment]
What do you think about defining less compare policy in convex_hull strategy?

**Comment by vissarion** [comment]
> Strategies are for CS-specific computation. AFAIU you want to compare coordinates strictly instead with epsilon in an algorithm. So the differentiation is made based on algorithm, not CS. This probably should not be a strategy but a policy (custom less policy) used only in convex_hull.
> 
> Consider this. The most general umbrella strategy can be passed to all algorithms, including `relate` and `convex_hull`. Which compare strategy should be returned by it?

@awulkiew that's a good point!

In https://github.com/boostorg/geometry/pull/1162/commits/f719a26b442eb12c0f91fba5cbb980758164cd77 I defined a new less policy (to be used only by convex hull for now) that passes the right equals policy to the strategy. Thus the strategy is responsible for the CS specific computation and the less policy (in the algorithm level) is responsible for the comparison of coordinates (exact or with epsilon). Alternatively, I could have added another template to less (i.e. EqualsPolicy). 

@awulkiew what do you think? I think it is better this way that defining less inside convex hull strategy, this way strategy does not have any information on how we compare coordinates but this is passed by the algorithm. Now type alias `compare` is the same in both `relate` and `convex hull` strategies.

BTW a similar problem holds with `side` method on those strategies, where `relate` uses a default side strategy while `convex hull` a robust one, this is a non-CS specific differentiation.

File: include/boost/geometry/algorithms/detail/relate/point_point.hpp (hunk 5 above)
Comment (barendgehrels) [reply include/boost/geometry/algorithms/detail/relate/point_point.hpp:225]: But `less` it is a public policy, meant for all library users. It’s not internal. So that we don’t use it like this does not mean that we can change it.

Comment (vissarion) [reply include/boost/geometry/algorithms/detail/relate/point_point.hpp:225]: This is updated and now the user can pass either a strategy or a cs tag as the third parameter to `less` (the second choice can be considered as deprecated and removed in the future). So in some sense now `less` is "enhanced". What do you think?

File: include/boost/geometry/algorithms/detail/disjoint/multipoint_geometry.hpp (hunk 2 above)
Comment (barendgehrels) [reply include/boost/geometry/algorithms/detail/disjoint/multipoint_geometry.hpp:68]: ok

File: include/boost/geometry/algorithms/detail/relate/point_point.hpp (hunk 5 above)
Comment (barendgehrels) [reply include/boost/geometry/algorithms/detail/relate/point_point.hpp:225]: sound good!

File: include/boost/geometry/policies/compare.hpp [hunk 7]
```diff
@@ -37,92 +38,244 @@ namespace boost { namespace geometry
...
+                strategy::compare::equals_epsilon
+            >::apply(left, right);
+    }
+};
+
+// for backward compatibility
+
+template <typename Point, int Dimension>
+struct less<Point, Dimension, boost::geometry::cartesian_tag>
+{
```
Comment (barendgehrels) [inline include/boost/geometry/policies/compare.hpp:128]: Ok, I see. Indeed `less` is enhanced now, and it wouldn’t break, and these are the implementations. 
So I’m ok now

**Review by barendgehrels — APPROVED** [review]
thanks!

**Comment by awulkiew** [comment]
Thanks, it looks better now. But I have one more concern. What if someone passes a different, non-robust or maybe non-cartesian side strategy (assuming it works) into the convex_hull? Would less_exact be the correct choice compatible with strategies other than robust side? Or am I missing something about the way how less compare is defined?

**Comment by vissarion** [comment]
So here the less strategy does have to do with robust side strategy. It is related to how we sort inside convex hull. If you sort with approximate less then you will run into issues like the one this PR solves when e.g. you have input points with very similar x or y coordinates. And these issues will be independent from side strategy.

**Comment by awulkiew** [comment]
Right, thanks for the explanation.
Since `less_exact` is explicitly only used in `convex_hull` does compare strategy has to be defined in umbrella strategies? AFAIU the only change that is needed is to use specific `less` in the algorithm. Or am I missing something?

**Comment by vissarion** [comment]
You mean to pass just the (non umbrella) compare strategy to `less_exact` or even define `less_exact` inside `convex_hull`? I implemented that way because I need to change how numbers are compared inside  compare strategies (e.g. https://github.com/boostorg/geometry/blob/develop/include/boost/geometry/strategies/compare.hpp#L93). Also, `less` and `less_exact` policies now take a strategy instead of a CStag which I think is more correct, right?

Moreover, even not tested, I think that there could be similar problems in the library where `std::sort` is used to sort points or turns. So, it is eventually not only convex hull to be affected.

**Comment by awulkiew** [comment]
Yes, I meant having `less_exact` policy alongside `less` policy, both taking CSTag and using `less_exact` in `convex_hull`. The rationale behind this was to introduce least amount of changes.

> take a strategy instead of a CStag which I think is more correct

To be honest I don't know which is more correct. I feel that the whole less policy could itself be a strategy because it's almost wholly CS-specific. But taking the strategy seems to be more "correct" than taking CSTag.

**Review by awulkiew — APPROVED** [review]
